File: ios/BockMedia/Features/Family/ProfilePickerGate.swift
import SwiftUI

/// Blocks until the user picks a household profile or explicitly continues unattributed.
struct ProfilePickerGate<Content: View>: View {
    @ObservedObject var appState: AppState
    @ViewBuilder var content: () -> Content

    @State private var members: [HouseholdMember]?
    @State private var loadFailed = false
    @State private var profileSyncing = false
    @State private var storeRevision = HouseholdStore.revision

    private var choiceMade: Bool { ActiveProfileStore.hasProfileChoice() }

    var body: some View {
        Group {
            if UITestSupport.isEnabled {
                content()
            } else if choiceMade && !profileSyncing && appState.shellReady {
                content()
            } else if !appState.shellReady || profileSyncing || (members == nil && !loadFailed) {
                LoadingBox()
            } else if loadFailed {
                pickerError
            } else if members?.isEmpty == true {
                content()
            } else {
                ZStack {
                    Color.black.opacity(0.55).ignoresSafeArea()
                    profilePickerDialog
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .task(id: "\(choiceMade)-\(storeRevision)-\(appState.shellReady)") {
            guard appState.shellReady else { return }
            guard !choiceMade else { return }
            await loadMembers()
        }
        .onReceive(NotificationCenter.default.publisher(for: HouseholdStore.changedNotification)) { _ in
            storeRevision = HouseholdStore.revision
            guard !choiceMade else { return }
            let fresh = HouseholdStore.members()
            if !fresh.isEmpty {
                members = fresh
                loadFailed = false
            }
        }
    }

    private var pickerError: some View {
        VStack(spacing: 12) {
            Text("Couldn't load family profiles")
                .font(.headline)
            Text(HouseholdStore.lastLoadError ?? "Check your connection and try again.")
                .font(.subheadline)
                .foregroundStyle(BockColors.muted)
                .multilineTextAlignment(.center)
            Button("Retry") {
                Task { await loadMembers() }
            }
            .foregroundStyle(BockColors.green)
            Button("Continue unattributed") {
                Task { await pickUnattributed() }
            }
            .font(.subheadline)
            .foregroundStyle(BockColors.muted)
        }
        .padding(24)
    }

    private func loadMembers() async {
        loadFailed = false
        if let cached = HouseholdStore.cached, !cached.members.isEmpty {
            members = cached.members
            return
        }
        if UITestSupport.isEnabled {
            members = []
            if !ActiveProfileStore.hasProfileChoice() {
                await pickUnattributed()
            }
            return
        }
        let fetched = await withAsyncTimeout(seconds: 12) {
            await HouseholdStore.refresh(repository: appState.repository, force: true)
        }
        switch fetched {
        case .success(let response):
            members = response.members
            loadFailed = response.members.isEmpty && HouseholdStore.lastLoadError != nil
        case .timedOut:
            let cached = HouseholdStore.members()
            if cached.isEmpty {
                members = nil
                loadFailed = true
            } else {
                members = cached
            }
        }
    }

    @ViewBuilder
    private var profilePickerDialog: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Who's listening?")
                .font(.headline)
            Text(
                "Pick a profile to restore your ratings and settings, " +
                "or continue unattributed until you choose later in Family."
            )
            .font(.subheadline)
            .foregroundStyle(BockColors.muted)
            ForEach(members ?? []) { member in
                Button {
                    Task { await pick(member) }
                } label: {
                    HStack {
                        VStack(alignment: .leading) {
                            Text(member.name)
                            Text(member.isParent ? "Parent" : "Kid")
                                .font(.caption)
                                .foregroundStyle(BockColors.muted)
                        }
                        Spacer()
                    }
                    .padding(.vertical, 8)
                }
                .buttonStyle(.plain)
                .disabled(profileSyncing)
                .accessibilityIdentifier(BockTestTags.profileMemberPick(member.id))
                Divider()
            }
            Button("Continue unattributed") {
                Task { await pickUnattributed() }
            }
            .frame(maxWidth: .infinity, alignment: .trailing)
            .foregroundStyle(BockColors.green)
            .disabled(profileSyncing)
            .accessibilityIdentifier(BockTestTags.profileContinueUnattributed)
        }
        .padding(20)
        .background(BockColors.sheetBg)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .padding(24)
        .accessibilityIdentifier(BockTestTags.profilePicker)
    }

    private func pick(_ member: HouseholdMember) async {
        profileSyncing = true
        defer { profileSyncing = false }
        let previous = ActiveProfileStore.activeMemberId()
        await ClientPrefsSync.onActiveMemberChanged(
            repository: appState.repository,
            memberId: member.id,
            previousMemberId: previous
        )
    }

    private func pickUnattributed() async {
        profileSyncing = true
        defer { profileSyncing = false }
        let previous = ActiveProfileStore.activeMemberId()
        await ClientPrefsSync.onActiveMemberChanged(
            repository: appState.repository,
            memberId: nil,
            previousMemberId: previous
        )
    }
}

File: android/app/src/main/kotlin/com/bockmedia/console/ui/family/ProfilePickerGate.kt
package com.bockmedia.console.ui.family

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.bockmedia.console.ui.testing.BockTestTags
import com.bockmedia.console.data.api.dto.HouseholdMember
import com.bockmedia.console.data.repository.BockMediaRepository
import com.bockmedia.console.local.ActiveProfileStore
import com.bockmedia.console.local.ClientPrefsSync
import com.bockmedia.console.local.HouseholdStore
import kotlinx.coroutines.launch

/** Blocks until the user picks a household profile or explicitly continues unattributed. */
@Composable
fun ProfilePickerGate(
    repository: BockMediaRepository,
    content: @Composable () -> Unit,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var members by remember { mutableStateOf<List<HouseholdMember>?>(null) }
    var loading by remember { mutableStateOf(false) }
    val choiceMade by ActiveProfileStore.profileChoiceMadeState.collectAsState()
    val householdRevision by HouseholdStore.revision.collectAsState()

    LaunchedEffect(choiceMade, householdRevision) {
        if (choiceMade) return@LaunchedEffect
        val cached = HouseholdStore.members()
        if (cached.isNotEmpty()) {
            members = cached
            return@LaunchedEffect
        }
        members = runCatching { repository.household().members }.getOrDefault(emptyList())
    }

    if (choiceMade) {
        content()
        return
    }

    when (members) {
        null -> {
            Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
        }
        emptyList<HouseholdMember>() -> content()
        else -> {
            AlertDialog(
                onDismissRequest = {},
                modifier = Modifier.testTag(BockTestTags.PROFILE_PICKER),
                title = { Text("Who's listening?") },
                text = {
                    Column(Modifier.fillMaxWidth()) {
                        Text(
                            "Pick a profile to restore your ratings and settings, " +
                                "or continue unattributed until you choose later in Family.",
                            style = MaterialTheme.typography.bodyMedium,
                        )
                        Spacer(Modifier.height(12.dp))
                        if (loading) {
                            Box(
                                Modifier.fillMaxWidth().padding(vertical = 8.dp),
                                contentAlignment = Alignment.Center,
                            ) {
                                CircularProgressIndicator()
                            }
                        }
                        LazyColumn(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                            items(members!!, key = { it.id }) { member ->
                                ListItem(
                                    headlineContent = { Text(member.name) },
                                    supportingContent = {
                                        Text(if (member.role == "parent") "Parent" else "Kid")
                                    },
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .clickable(enabled = !loading) {
                                            loading = true
                                            scope.launch {
                                                try {
                                                    ClientPrefsSync.onActiveMemberChanged(
                                                        context,
                                                        member.id,
                                                        ActiveProfileStore.activeMemberId(context),
                                                    )
                                                } finally {
                                                    loading = false
                                                }
                                            }
                                        },
                                )
                            }
                        }
                    }
                },
                confirmButton = {
                    TextButton(
                        enabled = !loading,
                        onClick = {
                            loading = true
                            scope.launch {
                                try {
                                    ClientPrefsSync.onActiveMemberChanged(
                                        context,
                                        null,
                                        ActiveProfileStore.activeMemberId(context),
                                    )
                                } finally {
                                    loading = false
                                }
                            }
                        },
                    ) {
                        Text("Continue unattributed")
                    }
                },
            )
        }
    }
}
